### PR TITLE
ICU-21266 Support toSkeleton() for all functional Unit Formatters

### PR DESCRIPTION
**Pull request by hugovdm**
Closes icu-units#35, closes icu-units#59.

##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21266
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [X] Tests included
- [X] Documentation is changed or added

### CODE REVIEW

**Review by younies — APPROVED** [review]
fix the travis and the building

**Review by younies — DISMISSED** [review]
Fix the travis and build

**Review by sffc — COMMENTED** [review]
Your code is good, but I don't know if I agree with the behavior.  I understand that you're trying to maintain backwards compatibility, but the result ends up being more confusing, especially if you're unfamiliar with the history.  I think I would prefer that we just output a normalized "unit/" stem in all cases.  We can discuss at the ICU-TC meeting this week.

File: icu4c/source/i18n/number_skeletons.cpp [hunk 1]
```diff
@@ -1520,14 +1526,33 @@ bool GeneratorHelpers::unit(const MacroProps& macros, UnicodeString& sb, UErrorC
...
+            // Combine perUnit with unit
+            MeasureUnitImpl temp;
+            const MeasureUnitImpl &perUnit =
+                MeasureUnitImpl::forMeasureUnit(macros.perUnit, temp, status);
+            for (int32_t i = 0; i < perUnit.units.length(); i++) {
+                const SingleUnitImpl *subUnit = perUnit.units[i];
+                if (subUnit->dimensionality > 0) {
+                    SingleUnitImpl newSub = *subUnit;
+                    newSub.dimensionality *= -1;
+                    unit = unit.product(newSub.build(status), status);
```
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_skeletons.cpp:1533]: Done.

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/number/NumberFormatterApiTest.java [hunk 2]
```diff
@@ -3392,6 +3515,27 @@ public void fieldPositionCoverage() {
                     expectedFieldPositions);
         }
 
+        {
+            String message = "Measure unit field position with prefix and suffix, built-in m/s";
+            FormattedNumber result = assertFormatSingle(
+                    message,
+                    "measure-unit/speed-meter-per-second unit-width-full-name",
+                    "unit/meter-per-second unit-width-full-name",
+                    NumberFormatter.with().unit(MeasureUnit.METER_PER_SECOND).unitWidth(UnitWidth.FULL_NAME),
```
Comment (hugovdm) [inline icu4j/main/tests/core/src/com/ibm/icu/dev/test/number/NumberFormatterApiTest.java:3524]: With this now so closely matching the "composed m/s" example above, I probably don't need both anymore and could collapse this into one again?

**Review by sffc — COMMENTED** [review]
I like the simplicity of the C++ changes. Are you planning to do the same in Java?

File: icu4c/source/i18n/number_formatimpl.cpp [hunk 3]
```diff
@@ -162,6 +162,20 @@ NumberFormatterImpl::macrosToMicroGenerator(const MacroProps& macros, bool safe,
     bool isMixedUnit = isCldrUnit && (uprv_strcmp(macros.unit.getType(), "") == 0) &&
                        macros.unit.getComplexity(status) == UMEASURE_UNIT_MIXED;
 
+    MeasureUnit unit = macros.unit;
+    MeasureUnit perUnit = macros.perUnit;
+    if (isCldrUnit && !utils::unitIsBaseUnit(perUnit)) {
+        // Simplify away perUnit for:
+        // * compoundUnit-per-anotherUnit, e.g. meter-per-second per second
+        // * the reust is a built-in unit, e.g. pound-force-per-square-inch
```
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_formatimpl.cpp]: I changed the comment, I think it should be an easier read now.

**Comment by hugovdm** [comment]
Changed this to ticket ICU-21266 since it addresses all of its concerns.

**Comment by hugovdm** [comment]
Travis failure might be a race between two copies of the same command producing the same output file:

```
pkgdata: clang -DU_ATTRIBUTE_DEPRECATED=   -DU_HAVE_STRTOD_L=1 -DU_HAVE_XLOCALE_H=1 -DU_HAVE_STRING_VIEW=1  -O2  -std=c11 -Wall -pedantic -Wshadow -Wpointer-arith -Wmissing-prototypes -Wwrite-strings  -Qunused-arguments -Wno-parentheses-equality -fno-common -c -I../../common -I../../common  -dynamic -o uconvmsg/uconvmsg_dat.o uconvmsg/uconvmsg_dat.S
pkgdata: clang -DU_ATTRIBUTE_DEPRECATED=   -DU_HAVE_STRTOD_L=1 -DU_HAVE_XLOCALE_H=1 -DU_HAVE_STRING_VIEW=1  -O2  -std=c11 -Wall -pedantic -Wshadow -Wpointer-arith -Wmissing-prototypes -Wwrite-strings  -Qunused-arguments -Wno-parentheses-equality -fno-common -c -I../../common -I../../common  -dynamic -o uconvmsg/uconvmsg_dat.o uconvmsg/uconvmsg_dat.S
pkgdata: ar r -c uconvmsg/libuconvmsg.a uconvmsg/uconvmsg_dat.o
pkgdata: ar r -c uconvmsg/libuconvmsg.a uconvmsg/uconvmsg_dat.o
pkgdata: ranlib uconvmsg/libuconvmsg.a
/Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: archive member: uconvmsg/libuconvmsg.a() size too large (archive member extends past the end of the file)
-- return status = 256
Error generating library file. Failed command: ranlib uconvmsg/libuconvmsg.a
```

File: icu4c/source/i18n/number_longnames.cpp [hunk 4]
```diff
@@ -223,31 +223,30 @@ void LongNameHandler::forMeasureUnit(const Locale &loc, const MeasureUnit &unitR
...
-        // error code. Once we support not-built-in units here, unitRef may be
-        // anything, but if not built-in, perUnit has to be "none".
-        status = U_UNSUPPORTED_ERROR;
-        return;
-    }
 
     MeasureUnit unit = unitRef;
     if (uprv_strcmp(perUnit.getType(), "none") != 0) {
-        // Compound unit: first try to simplify (e.g. "meter per second" is a
-        // built-in unit).
```
Comment (sffc) [inline icu4c/source/i18n/number_longnames.cpp:239]: You deleted this comment in C++ but not in Java.  Probably keep it in both.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/units/MeasureUnitImpl.java [hunk 5]
```diff
@@ -70,7 +70,9 @@ public MeasureUnitImpl copy() {
         MeasureUnitImpl result = new MeasureUnitImpl();
         result.complexity = this.complexity;
         result.identifier = this.identifier;
-        result.singleUnits = new ArrayList<>(this.singleUnits);
+        for (SingleUnitImpl single : this.singleUnits) {
```
Comment (sffc) [inline icu4j/main/classes/core/src/com/ibm/icu/impl/units/MeasureUnitImpl.java:73]: Good catch!  I guess this is a Java-only change?

**Review by sffc — APPROVED** [review]

**Comment by sffc** [comment]
The Travis mac build is erroring out with:

```
pkgdata: ranlib uconvmsg/libuconvmsg.a
/Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: archive member: uconvmsg/libuconvmsg.a() size too large (archive member extends past the end of the file)
-- return status = 256
Error generating library file. Failed command: ranlib uconvmsg/libuconvmsg.a
Error generating assembly code for data.
```

Probably a flake.  @richgillam?

**Comment by richgillam** [comment]
>The Travis mac build is erroring out... 
>
>Probably a flake. @richgillam?

Why are you asking _me_?  I don't think I know anything about what's going on here.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/units/MeasureUnitImpl.java (hunk 5 above)
Comment (hugovdm) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/units/MeasureUnitImpl.java:73]: In C++, MeasureUnitImpl::copy() has this in a for-loop, making use of the copy-constructor:

    SingleUnitImpl *item = result.units.emplaceBack(*units[i]);

File: icu4c/source/i18n/number_longnames.cpp (hunk 4 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:239]: Done.

**Comment by hugovdm** [comment]
Whoops: 1379 introduced a few more unit tests that need skeletons now.

**Comment by sffc** [comment]
> > The Travis mac build is erroring out...
> > Probably a flake. @richgillam?
> 
> Why are you asking _me_? I don't think I know anything about what's going on here.

I was asking you because it was a Mac build failure and I thought you might know more what it meant.

**Comment by richgillam** [comment]
>I was asking you because it was a Mac build failure and I thought you might know more what it meant.

Thank you; I didn't pick up on that.  No, I have no idea what might be going on there.  If it's not a one-time fluke and we need this, I can try to figure it out, but I bring no specific knowledge to bear on that, so it might take me a while.

**Comment by hugovdm** [comment]
> > I was asking you because it was a Mac build failure and I thought you might know more what it meant.
> 
> Thank you; I didn't pick up on that. No, I have no idea what might be going on there. If it's not a one-time fluke and we need this, I can try to figure it out, but I bring no specific knowledge to bear on that, so it might take me a while.

I copy-pasted the error with some more context above - here it is again:

```
pkgdata: clang -DU_ATTRIBUTE_DEPRECATED=   -DU_HAVE_STRTOD_L=1 -DU_HAVE_XLOCALE_H=1 -DU_HAVE_STRING_VIEW=1  -O2  -std=c11 -Wall -pedantic -Wshadow -Wpointer-arith -Wmissing-prototypes -Wwrite-strings  -Qunused-arguments -Wno-parentheses-equality -fno-common -c -I../../common -I../../common  -dynamic -o uconvmsg/uconvmsg_dat.o uconvmsg/uconvmsg_dat.S
pkgdata: clang -DU_ATTRIBUTE_DEPRECATED=   -DU_HAVE_STRTOD_L=1 -DU_HAVE_XLOCALE_H=1 -DU_HAVE_STRING_VIEW=1  -O2  -std=c11 -Wall -pedantic -Wshadow -Wpointer-arith -Wmissing-prototypes -Wwrite-strings  -Qunused-arguments -Wno-parentheses-equality -fno-common -c -I../../common -I../../common  -dynamic -o uconvmsg/uconvmsg_dat.o uconvmsg/uconvmsg_dat.S
pkgdata: ar r -c uconvmsg/libuconvmsg.a uconvmsg/uconvmsg_dat.o
pkgdata: ar r -c uconvmsg/libuconvmsg.a uconvmsg/uconvmsg_dat.o
pkgdata: ranlib uconvmsg/libuconvmsg.a
/Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: archive member: uconvmsg/libuconvmsg.a() size too large (archive member extends past the end of the file)
-- return status = 256
Error generating library file. Failed command: ranlib uconvmsg/libuconvmsg.a
```

To me it looks like it's executing the same command twice, overwriting the same file twice, making me wonder if there could be a race in writing and reading uconvmsg/uconvmsg_dat.o ?

**Comment by hugovdm** [comment]
Commit 422f3de5935eb7ee3d650c4e71d3c241717da907 is new - I've "unsquashed", will resquash as soon as _some_ tests pass, letting the tests run on the squashed version and hopefully awaiting only one final approval.

**Comment by richgillam** [comment]
>To me it looks like it's executing the same command twice, overwriting the same file twice, making me wonder if there could be a race in writing and reading uconvmsg/uconvmsg_dat.o and/or uconvmsg/libuconvmsg.a ?

It looks that way to me too, and that seems like a plausible hypothesis.

**Review by sffc — APPROVED** [review]

**Pull request merged**